File: AdjoinableMPIexamples/Tapenade_F77/SendIrecvWaitTwoWay/programAMPI_d.f
c   DESIRED TAPENADE TANGENT USING THE AMPI LIBRARY
c    Simple Irecv+Send+Wait || Irecv+Wait+Send

      SUBROUTINE head_d(x,xd,y,yd)
      implicit none
      include 'ampi/ampif.h'
      DOUBLE PRECISION x,y
      DOUBLE PRECISION xd,yd
      INTEGER world_rank, err_code, req
      DOUBLE PRECISION local
      DOUBLE PRECISION locald

      call AMPI_Comm_rank(MPI_COMM_WORLD, world_rank, err_code)
      IF (world_rank.EQ.0) THEN
         call TLS_AMPI_IRecv(y, yd, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 1, 0,
     +        AMPI_FROM_SEND, MPI_COMM_WORLD, req, err_code)
         xd=xd*2
         x=x*2
         call TLS_AMPI_Send(x, xd, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 1, 0,
     +        AMPI_TO_IRECV_WAIT, MPI_COMM_WORLD, err_code)
         call TLS_AMPI_Wait(req, MPI_STATUS_IGNORE, err_code)
         yd=yd*3
         y=y*3
      ELSE IF (world_rank.EQ.1) THEN
         call TLS_AMPI_IRecv(local, locald, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 0, 0,
     +        AMPI_FROM_SEND, MPI_COMM_WORLD, req, err_code)
         call TLS_AMPI_Wait(req, MPI_STATUS_IGNORE, err_code)
         locald = COS(local)*locald
         local = SIN(local)
         call TLS_AMPI_Send(local, locald, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 0, 0,
     +        AMPI_TO_IRECV_WAIT, MPI_COMM_WORLD, err_code)
      ENDIF
      END

      PROGRAM main
      implicit none
      include 'ampi/ampif.h'
      INTEGER world_rank, err_code
      DOUBLE PRECISION x,y
      DOUBLE PRECISION xd,yd

      call AMPI_Init_NT(err_code)
      call AMPI_Comm_rank(MPI_COMM_WORLD, world_rank, err_code)
      IF (world_rank.EQ.0) THEN
         xd=1.0
         x=2.5
         print *,"process",world_rank," sends val  [ 2.500000000==]",x
         call head_d(x,xd,y,yd)
         print *,"process",world_rank," recvs val  [-2.876772823==]",y
         print *,"process",world_rank," recvs diff [1.7019731128==]",yd
         yd=yd+xd
         y=y+x
      ELSE IF (world_rank.EQ.1) THEN
         call head_d(x,xd,y,yd)
      ENDIF
      call AMPI_Finalize_NT(err_code)
      END

File: AdjoinableMPIexamples/Tapenade_C/SendIrecvWaitTwoWay/programAMPI_d.c
/*  DESIRED TAPENADE TANGENT USING THE AMPI LIBRARY */
/* Simple Irecv+Send+Wait || Irecv+Wait+Send */

#include <stdio.h>
#include <math.h>
#include "ampi/ampi.h"

void head_d(double *x, double *xd, double *y, double *yd) {
  AMPI_Request r;
  int world_rank;
  AMPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  if (world_rank == 0) {
    TLS_AMPI_Irecv(y, yd, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 1, 0,
                   AMPI_FROM_SEND, MPI_COMM_WORLD, &r);
    *xd=*xd*2;
    *x=*x*2;
    TLS_AMPI_Send(x, xd, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 1, 0,
                  AMPI_TO_IRECV_WAIT, MPI_COMM_WORLD);
    TLS_AMPI_Wait(&r,MPI_STATUS_IGNORE);
    *yd=*yd*3;
    *y=*y*3;
  } else if (world_rank == 1) {
    double local;
    double locald;
    TLS_AMPI_Irecv(&local, &locald, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 0, 0,
                   AMPI_FROM_SEND, MPI_COMM_WORLD, &r);
    TLS_AMPI_Wait(&r,MPI_STATUS_IGNORE);
    locald=cos(local)*locald ;
    local=sin(local);
    TLS_AMPI_Send(&local, &locald, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 0, 0,
                  AMPI_TO_IRECV_WAIT, MPI_COMM_WORLD);
  }
}

int main(int argc, char** argv) {
  AMPI_Init_NT(0,0);
  int world_rank;
  AMPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double x,y;
  double xd,yd;
  if (world_rank == 0) {
    xd=1.0;
    x=2.5;
    printf(__FILE__ ": process %i sends val [2.500000==]%f\n", world_rank, x);
    head_d(&x,&xd,&y,&yd);
    printf(__FILE__ ": process %i recvs val [-2.876773==]%f\n", world_rank, y);
    printf(__FILE__ ": process %i recvs diff [1.701973==]%f\n", world_rank, yd);
    y=y+x ;
  }else if (world_rank == 1) {
    head_d(&x,&xd,&y,&yd);
  }
  AMPI_Finalize_NT();
  return 0;
}
